### PR TITLE
formatting update

### PR DESCRIPTION
**Pull request by kjmerf**
submitting meaningless pull request to test process

### CODE REVIEW

File: datascroller/scroller.py
Comment (johncmerfeld) [inline datascroller/scroller.py:86]: This is an example of a comment I might write. Nice!

**Review by baogorek — DISMISSED** [review]
Thanks for the style fixes, @kjmerf!

**Comment by baogorek** [comment]
@kjmerf, will you do one more thing for me? Since this is the first change after version 1.1.0, I was wondering what we should do with the version number. Then I came across [this post](https://softwareengineering.stackexchange.com/questions/166215/when-do-you-change-your-major-minor-patch-version-number) that argues that one should immediately update the version number to the one we're working *towards*. @johncmerfeld , interested to hear your take on this. What I think that means is that this pull request should also modify setup.py to have version 1.2.0, which is the next minor release. Let me know if that makes sense.

**Review by johncmerfeld — COMMENTED** [review]
> @kjmerf, will you do one more thing for me? Since this is the first change after version 1.1.0, I was wondering what we should do with the version number. Then I came across [this post](https://softwareengineering.stackexchange.com/questions/166215/when-do-you-change-your-major-minor-patch-version-number) that argues that one should immediately update the version number to the one we're working _towards_. @johncmerfeld , interested to hear your take on this. What I think that means is that this pull request should also modify setup.py to have version 1.2.0, which is the next minor release. Let me know if that makes sense.

Check out the original [semantic versioning](https://semver.org) site. Quick read and interesting. Right at the top the specification says, "Once a versioned package has been released, the contents of that version MUST NOT be modified. Any modifications MUST be released as a new version." I think this is a good rule of thumb. Since the goal with this upcoming development is to introduce **new, backwards-compatible** functionality, I'd recommend upping to 1.2.0, and then incrementing the patch (third) number as we merge code. Nothing wrong with releasing a version like 1.2.4. Now that this is a potentially public dependency, we should be good stewards of it!

**Comment by kjmerf** [comment]
@johncmerfeld are you suggesting we include an update to setup.py in each pull request? Seems like that could be problematic if we have multiple people working at once... I foresee merge conflicts

**Comment by johncmerfeld** [comment]
> @johncmerfeld are you suggesting we include an update to setup.py in each pull request? Seems like that could be problematic if we have multiple people working at once... I foresee merge conflicts

I am suggesting that, yeah. Merge conflicts aren't inherently problematic, IMO. If anything, the act of resolving those conflicts will impose an absolute order on updates to the package, which is desirable.

**Comment by kjmerf** [comment]
Fine with me. Going to close this PR and open new one regarding formatting.

**Comment by kjmerf** [comment]
Fine with me. Also, I added a couple issues to the backlog after playing
around a bit. Curious to hear your guys thoughts!

Alright, back to the day job....

**Comment by baogorek** [comment]
@kjmerf, I'm going to open this back up; I definitely want your style updates to get merged. Would also have been fine with you adding the setup.py change in this pull request as well but if that change gets merged in another PR first that's fine too. (As you've probably noticed, you can push new files to a branch with a pull request and they will show up in the pull request.)

@johncmerfeld, let's discuss version number strategy a bit more on this PR. Remembering the good old days at our former employer, we made practical use of the increment version number every commit rule. Since the push to that employer's CRAN-like repository happened all the time, there was a very real risk of two people having a package with the same version number but different code. I think we did the right thing there.

Here I feel things *might* be different, since we're in complete control of when we release to PiPy. My thought is, we could have concrete plans for minor version 1.2.0 but notice something in version 1.1.0 that needs fixing, and simultaneously be working on branches with the version set to 1.1.1 and 1.2.0. In that case two people could be working on a branch with minor version 1.2.0, and as long as the plan was being followed for that minor version, they could both have the same one.

Or we could keep incrementing like we did. If we do that, there's one other thing we need to figure out: what do we do with the download_url in setup.py. I'd rather not create a release for every commit. I guess it wouldn't be the end of the world to point to an old release's download URL. Looking forward to hearing your thoughts.

Ben

**Comment by kjmerf** [comment]
@baogorek if you want to keep this PR open for sake of versioning discussion that's fine but I moved the formatting changes to the other PR I have open (#19). Let's merge that one if all looks ok

**Comment by baogorek** [comment]
@kjmerf sounds good. Let's keep it open just long enough for the discussion.

**Comment by kjmerf** [comment]
Cool. I'll wait for @johncmerfeld to reply because looks like we merged the data type fix (#22) without updating the setup.py version. I'm not familiar with pypi so I'm going to take a backseat in this discussion.

**Comment by johncmerfeld** [comment]
@baogorek Okay, good points. I've never worked on software that actually gets distributed publicly; definitely warrants thinking about a different model. But I'm not crazy about working on 1.2 and 1.1.1 at the same time, for example. We want to a) have a consistent versioning scheme, b) not have to do a release every commit, and c) not overthink it. Here's my proposed model:

  0. Let's say we just released version 1.1. 
  1. Initially, we decide what will be in version 1.2. Might be a set of issues or features or whatever. 
  2. We start writing code, and we **don't pay heed to version numbers.** i.e. the onus to edit setup.py is never on the author while they're editing.
  3. During the pull request process, we decide whether what's being merged warrants a new release.
  4. If it warrants a release, we increment to 1.2.0. The next time, we increment to 1.2.1. Let's say we get up to version 1.2.4. **If we've accomplished all the goals we set for 1.2, we agree that 1.2.4 will be the last version prefixed with 1.2**, and we announce it as "version 1.2." Then we repeat the process from step 1, with goals for 1.3.
  5. If it doesn't warrant a release, we just merge it. No change to the version number, no new release. 

I like the intuitions created with this way of doing it. If someone knows they need a historical version like 1.2, they can see that the version with 1.2's features and all the bug fixes and stuff is whichever one has the highest patch number (1.2.4). But people can also pull the latest version as soon as we fix something important.

The only issue here is that we could potentially have people downloading code directly from the repo with a version number that matches what's in PyPi but with different code. That's when you start getting into build numbers and stuff but, I think that violates "don't overthink it."

This took me longer to think about than I hoped it would, and maybe it's still overthinking the problem. Curious to hear from both of you.

**Comment by johncmerfeld** [comment]
As part of the above, we could open dedicated issues to discuss upcoming minor versions or major versions. For example, @kjmerf's database connection idea would, I think, warrant a 2.0 release. (But we could also reserve that for non-backwards compatible stuff). And we could use the issue as a forum to sketch out what specific requirements we need before we decide it's ready. Hm… but does this contradict the "increment immediately" idea? I guess, in the above example, once we implemented the first "major version requirement" (i.e. very first database connection capability), we increment to 2.0.0. Then the "2.0 release" would end up being version 2.0.2 or whatever it was.

I dunno man, maybe I'm down a total rabbit hole. This makes sense to me, but what do y'all think?

**Comment by baogorek** [comment]
@johncmerfeld I think your policy is a good middle ground between incrementing the version every PR and planning out every release ahead of time or developing on multiple versions simultaneously (a la python 2 and python 3). I'm fine with letting pull requests be the mechanism for deciding whether we increment. We could lso err on the side of incrementing more frequently than necessary.

About the "increment immediately" idea, how about this: as part of the PyPi release process, right after the code lands on PyPi, we do a one-line PR updating the patch number in setup.py. PyPi's version will be immediately behind the master version, but I think that's okay. If we end up releasing to PyPi after every pull request, then we just ignore the suggestion. There are people that argue against the methodology from the Semantic Versioning webpage.

Sound good to you? If so, we'll call it a working policy and close this PR.

**Comment by johncmerfeld** [comment]
@baogorek I like it! Let's call it a policy.

**Comment by baogorek** [comment]
Policy set! Closing this PR turned discussion thread.

**Pull request closed without merging**
